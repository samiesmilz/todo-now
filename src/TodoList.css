.TodoList {
  background-color: antiquewhite;
  color: black;
  padding: 10px;
}
.TodoList-items {
  font-size: small;
  padding-left: 15px;
}
