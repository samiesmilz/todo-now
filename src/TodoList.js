import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import "./TodoList.css";

const TodoList = () => {
  const [todos, setTodos] = useState([]);

  const addTodo = ({ id, todo }) => {
    setTodos(() => [...todos, { id, todo }]);
  };

  const remove = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };

  const update = (id, newData) => {
    setTodos(
      todos.map((item) => (item.id === id ? { ...item, todo: newData } : item))
    );
  };

  const todoItem = todos.map((todo) => (
    <Todo
      item={todo.todo}
      remove={remove}
      id={todo.id}
      key={todo.id}
      update={update}
    />
  ));

  return (
    <div className="TodoList">
      <NewTodoForm addTodo={addTodo} />
      <ol className="TodoList-items">{todoItem}</ol>
    </div>
  );
};

export default TodoList;
