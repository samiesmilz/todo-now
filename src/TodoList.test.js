import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

const addNewTodo = (todo = "Test this todo app") => {
  // Select input elements and the submit button
  const todoInput = screen.getByPlaceholderText("Enter todo");
  const addTodoButton = screen.getByText("Add todo");

  // Enter values in the input elements
  fireEvent.change(todoInput, { target: { value: todo } });

  // fire an event to submit
  fireEvent.click(addTodoButton);
};

it("Adds a new todo on submit", () => {
  render(<TodoList />);
  // Check to make sure there are no boxes on the page.
  expect(screen.queryByText("remove")).not.toBeInTheDocument();
  expect(screen.queryByText("edit")).not.toBeInTheDocument();

  // Add a new box and expect to see this text now.
  addNewTodo();
  expect(screen.getByText("remove")).toBeInTheDocument();
  expect(screen.getByText("edit")).toBeInTheDocument();
});

it("can edit a todo", () => {
  render(<TodoList />);
  addNewTodo();

  fireEvent.click(screen.getByText("edit"));
  expect(screen.getByText("Update!")).toBeInTheDocument();

  const editInput = screen.getByDisplayValue("Test this todo app");
  fireEvent.change(editInput, { target: { value: "Working well" } });
  fireEvent.click(screen.getByText("Update!"));

  // expect only edited todo to appear
  expect(screen.getByText("Working well")).toBeInTheDocument();
  expect(screen.queryByText("Test this todo app")).not.toBeInTheDocument();
});

it("can remove a todo", () => {
  render(<TodoList />);
  addNewTodo();
  const removeButton = screen.getByText("remove");
  fireEvent.click(removeButton);
  expect(screen.queryByText("remove")).not.toBeInTheDocument();
  expect(screen.queryByText("edit")).not.toBeInTheDocument();
});
