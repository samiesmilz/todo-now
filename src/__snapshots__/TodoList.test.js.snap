// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="NewTodoForm"
    >
      <form>
        <label>
          Todo
        </label>
        <input
          class="NewTodoForm-input"
          id="todo"
          name="todo"
          placeholder="Enter todo"
          type="text"
          value=""
        />
        <button
          class="NewTodoForm-button"
          type="submit"
        >
          Add todo
        </button>
      </form>
    </div>
    <ol
      class="TodoList-items"
    />
  </div>
</DocumentFragment>
`;
